### DISCUSSION TITLE
How to apply mask for all the models?

### DISCUSSION THREAD

**Discussion by NahidHasan09** [cvnlab/GLMsingle#80] [Q&A]
Hello,
To save time, I applied a mask by changing the default parameter 'hrrfitmask' from 1 to an array (a mask image) of 0s and 1s with the same shape as the data. However, it's still taking the same time to fit Model B. Is there any way to fit the models to specific voxels rather than the whole brain? Any suggestion would be highly appreciated.
Thanks!

**Comment by kendrickkay** [comment]
Hi - In general, if computational time is a concern, then one route would be to "preselect" voxels and make the <data> variable smaller prior to calling GLMsingle.  So, like, instead of feeding in all voxels (e.g. X x Y x Z x TR), you could mask/select and instead feed in only some voxels (e.g. voxels x TR).  You would have to keep track, of course, of how you did this selection process.  Does that make sense?

**Reply by NahidHasan09** [reply]
Thanks, I just tried this method. I preselected the voxels using nilearn's NifiMasker.transform_single_imgs function. However, the beta images looks bizarre (not comparable with our previous analysis of the same data on SPM). For the data in voxels x TR format, do I have to optimize the design matrices differently ( in my case they have same number of runs and time points as the data)?

**Reply by kendrickkay** [reply]
no, nothing special is needed. however, if the selection of voxels is very limited, there might be some strange issues. another thing you could quickly try is just select one slice and run on one slice. can you show pictures of your voxel mask or provide some other diagnostic information ?

**Reply by NahidHasan09** [reply]
Hi, I tried on one slice but the output doesn't look satisfying. In my experiment, the TRs, and onsets were not aligned, so I upsampled the data (using pyslicetime) and design matrix 3 times. I created the design matrix with glmsingle's make design matrix function and matched it to the data's shape. Actually, I am not sure whether I am creating the design matrix appropriately. So, I attached my code and onset files from two runs. Could you please check if I am doing it right or not? I have also attached the mask image and output from single slice run. Thanks!
[Diagnostic_glmsingle.zip](https://github.com/cvnlab/GLMsingle/files/9447082/Diagnostic_glmsingle.zip)

**Comment by kendrickkay** [comment]
![Single_Slice_TypeD](https://user-images.githubusercontent.com/2728338/187314089-ec784ef8-2b5c-416e-add9-dedc3158340e.png)

Hi - It is difficult for me to troubleshoot your event times, without a detailed description...  However, looking at Single_Slice_TypeD.png, I can tell a few things.  The R2 and FRACvalue results suggest that things are "sane".  There is a clear signal.  But it is highly localized.  One thing to keep in mind is that just because some voxels don't look "great" in the R2 and FRACvalue results doesn't mean that there aren't good experimental signals present.  We plan to update some examples with thoughts along these lines.  In the meantime, you could take the single-trial betas that are getting output and do further analyses.  It may be that the data are fine.  

There is a deeper issue here, which is whether your current experiment can be improved in terms of its general SNR.  That is a separate discussion, and gets into issues like experimental design and things like that.  

But from what I see, things seem to be sane.
